nemo/lightning/pytorch/plugins/precision.py

```diff
@@ -26,19 +27,112 @@
 AnyT = TypeVar("AnyT")
 
 
-class MegatronMixedPrecision(MixedPrecision):
+@dataclass
+class DtypeConfig:
+    fp32: bool = False
+    fp16: bool = False
+    bf16: bool = False
+    params_dtype: torch.dtype = None
+    pipeline_dtype: torch.dtype = None
+    autocast_dtype: torch.dtype = None
+    autocast_enabled: bool = False
+    grad_reduce_in_fp32: bool = True
+    # fp8 related
+    fp8: str = None
+    fp8_margin: int = 0
+    fp8_interval: int = 1
+    fp8_amax_history_len: int = 1
+    fp8_amax_compute_algo: str = "most_recent"
+    fp8_wgrad: bool = True
+    fp8_dot_product_attention: bool = False
+    fp8_multi_head_attention: bool = False
+
+
+def make_precision_from_dtypeconfig(config):
+    if config.fp32:
+        return "fp32"
+    elif config.fp16:
+        return 'fp16-mixed'
+    elif config.bf16:
+        return 'bf16-mixed'
+    else:
+        raise ValueError("Expected FP32, BF16 or FP16 to be true")
+
+
+def make_default_dtype_config_from_precision(precision: str):
+    """
+    Takes as input a precision string (or int) and returns a DtypeConfig with some default values.
+    TODO(@akoumparouli): defaults subject to change.
+    """
+    # cast to str if it's int
+    if isinstance(precision, int):
+        precision = str(precision)
+    is_bf16 = lambda x: x in ['bf16', 'bf16-mixed']
+    is_fp16 = lambda x: x in ['fp16', 'fp16-mixed', '16']
+    is_fp32 = lambda x: x in ['fp32', '32']
+
+    dtype = torch.bfloat16 if is_bf16(precision) else torch.float32
+    config = DtypeConfig(
+        fp32=is_fp32(precision),
+        fp16=is_fp16(precision),
+        bf16=is_bf16(precision),
+        params_dtype=torch.bfloat16,
+        pipeline_dtype=dtype,
+        autocast_dtype=dtype,
+        autocast_enabled=False,  #  for now
+        grad_reduce_in_fp32=True,
+    )
+    assert (
+        config.fp32 or config.fp16 or config.bf16
+    ), f"Error: Expected precision to be FP32, BF16 or FP16, {precision}"
+    return asdict(config)
+
+
+class MegatronCustomPrecision(MixedPrecision):
     def __init__(
         self,
-        precision: Literal["16-mixed", "bf16-mixed"],
+        fp32: bool = False,
+        fp16: bool = False,
+        bf16: bool = False,
+        params_dtype: torch.dtype = None,
+        pipeline_dtype: torch.dtype = None,
```

Comment:
Is there ever a situation where you want this in a different precision as the params?

Reply:
what if a user wants to initialize their model with FP32 but run training on bf16?

I'm not exactly sure it's unnecessary, but if that's the case I would prefer to have it removed from mcore first and reflect the change on NeMo as well.